aten/src/ATen/core/jit_type.h

```diff
@@ -24,6 +24,17 @@ struct Function;
 
 namespace c10 {
 
+// Better type comparison, since `operator=` doesn't work if the
+// pointers are different. We compare on Kind (fast), then compare by
+// the serialized representation (slower, but guaranteed to break ties).
+// This is a struct so that it can be used in certain std container
+// types.
+struct TypeEqual {
```

Comment:
this isn't correct for Unions/Optionals. In general we should be wary of code that attempts to optimize for negligible perf gains but adds complexity and maintenance to the system